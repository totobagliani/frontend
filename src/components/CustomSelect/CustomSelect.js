import React from 'react';
import { selectProductsByFilter } from '../../redux/selectors';

export default function CustomSelect({ selectProps, handleChange = () => {} }) {
  /*  const selectProp = {
    selectName: 'A name',
    classSelect: 'choose',
    selectTitle: 'Choose a section',
    classTitle: 'Classname Title',
    classOption: 'classname for the opcion"
    optionValues: []
  };   PropTypes */

  return (
    <select
      name={selectProps.selectName}
      className={selectProps.classSelect}
      defaultValue=""
      onChange={handleChange}
    >
      <option value="" className={selectProps.classTitle}>
        {selectProps.selectTitle}
      </option>
      {selectProps.optionValues.map((value) => (
        <option
          key={`${value}-id`}
          value={value}
          className={selectProps.classOption}
        >
          {value} - {selectProductsByFilter(value).length}
        </option>
      ))}
    </select>
  );
}
