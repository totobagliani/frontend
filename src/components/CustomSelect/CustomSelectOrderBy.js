/* eslint-disable comma-dangle */
import React from 'react';

export default function CustomSelectOrderBy({
  selectProps,
  handleChange = () => {}
}) {
  return (
    <select
      name={selectProps.selectName}
      className={selectProps.classSelect}
      defaultValue=""
      onChange={handleChange}
    >
      <option value="" className={selectProps.classTitle}>
        {selectProps.selectTitle}:
      </option>
      {selectProps.optionValues.map((value) => (
        <option
          key={`${value}-id`}
          value={value}
          className={selectProps.classOption}
        >
          {value}
        </option>
      ))}
    </select>
  );
}
