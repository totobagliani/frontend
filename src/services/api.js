/* eslint-disable comma-dangle */
export const getAllProducts = async () => {
  const baseURL = process.env.REACT_APP_API_URL;
  const resourcePath = 'api/products';
  try {
    const resp = await fetch(`${baseURL}/${resourcePath}/`);
    const data = await resp.json();
    return data;
  } catch (error) {
    return error;
  }
};

export const searchProductsByTitleTerm = async (term) => {
  const baseURL = process.env.REACT_APP_API_URL;
  const resourcePath = 'api/products';
  const queryParam = 'search?term';
  try {
    const resp = await fetch(
      `${baseURL}/${resourcePath}/${queryParam}=${term}`
    );
    const data = await resp.json();
    return data;
  } catch (error) {
    return error;
  }
};

export const addProduct = async (product) => {
  // localhost:5000/api/products/  -POST
  const baseURL = process.env.REACT_APP_API_URL;
  const resourcePath = 'api/products';

  try {
    const resp = await fetch(`${baseURL}/${resourcePath}/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json;charset=utf-8'
      },
      body: JSON.stringify(product)
    });
    const result = await resp.json();
    return result;
  } catch (error) {
    return error;
  }
};
