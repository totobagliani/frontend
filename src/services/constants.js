export const SECTIONS = [
  'baño',
  'dormitorio',
  'cocina',
  'salon',
  'terraza',
  'comedor',
];
export const SECTIONSOBJ = {
  BAÑO: 'baño',
  DORMITORIO: 'dormitorio',
  COCINA: 'cocina',
  SALON: 'salon',
  TERRAZA: 'terraza',
  COMEDOR: 'comedor',
};

// map. Object.values
export const SORT_OPTIONS = ['precio', 'nombre', 'seccion'];
