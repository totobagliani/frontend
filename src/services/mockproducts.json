{
  "ok": true,
  "products": [
    {
      "productName": "Mesa de baño",
      "description": "mesa de madera linda",
      "imageURL": "https://assets.leroymerlin.es/is/image/lmes/15769985-0400/estanteria-2-estantes-lotus-madera-27x70.jpg?$lmesBgMobile$&fit=constrain,0",
      "isFavourite": false,
      "price": 85.99,
      "section": "baño",
      "__v": 0,
      "id": "6298d964955b7b7648a8c413"
    },
    {
      "productName": "jarra",
      "description": "jarra linda",
      "imageURL": "https://assets.leroymerlin.es/is/image/lmes/15769985-0400/estanteria-2-estantes-lotus-madera-27x70.jpg?$lmesBgMobile$&fit=constrain,0",
      "isFavourite": false,
      "price": 15.99,
      "section": "baño",
      "__v": 0,
      "id": "6298de51913731ca34285593"
    },
    {
      "productName": "Set de Baño",
      "description": "Set Baño",
      "imageURL": "https://www.codisbath.com/wp-content/uploads/2021/02/mugarri-entrevista-codis-27.jpg",
      "isFavourite": false,
      "price": 500,
      "section": "baño",
      "id": "629e1301e655592b12cbb424"
    },

    {
      "productName": "Dormitorio",
      "description": "01 Dormitorio ",
      "imageURL": "https://dismobel.es/44535-large_default/dormitorio-boho-20-de-lan-mobel.jpg",
      "isFavourite": true,
      "price": 99.9,
      "section": "dormitorio",
      "id": "629e1531e655592b12cf5362"
    },

    {
      "productName": "Juego para cocina",
      "description": "Juego para cocina completo. ",
      "imageURL": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/0181-c2101-r1529506-1531506680.jpg",
      "isFavourite": true,
      "price": 99.9,
      "section": "cocina",
      "id": "629e16afe655592b12d1cf83"
    },

    {
      "productName": "Salón red",
      "description": "Salón 01 red",
      "imageURL": "https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/salon-moderno-blanco-con-vigas-1600353658.jpg",
      "isFavourite": false,
      "price": 9799.9,
      "section": "salon",
      "id": "629e1807e655592b12d41852"
    }
  ]
}
