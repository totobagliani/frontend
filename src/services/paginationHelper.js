const products = Array.from(new Array(50))
  .fill(0)
  .map((_, index) => index + 1);

const ELEMENTS_PER_PAGE = 6;

// let initialPageIndex = 0
// let finalPageIndex= 0

const currentPage = 1; // const [currentPage, setCurrentPage] = useState(1) => usePagination()

function paginate(array, page_size, page_number) {
  // human-readable page numbers usually start with 1, so we reduce 1 in the first argument
  console.log((page_number - 1) * page_size);

  const initialPageIndex = (page_number - 1) * page_size;
  const finalPageIndex = page_number * page_size;

  return array.slice(initialPageIndex, finalPageIndex);
}

// ProductsList
// useEffect(() => { const partialProductsByCurrentPage = paginate(products, ELEMENTS_PER_PAGE, currentPage) }, [currentPage])
paginate(products, ELEMENTS_PER_PAGE, currentPage);
