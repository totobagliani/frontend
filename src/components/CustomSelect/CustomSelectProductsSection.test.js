/* eslint-disable linebreak-style */
/* eslint-disable comma-dangle */
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { SORT_OPTIONS } from '../../services/constants';
import configureStore from '../../redux/store';
import CustomSelect from './CustomSelectProductsSection';

const initialState = {
  products: {
    products: [],
    productAdded: {
      productName: '',
      description: '',
      imageURL: '',
      isFavourite: false,
      price: 0,
      section: ''
    },
    searchResults: {
      itemsCount: 0,
      searchTerm: ''
    }
  },
  ui: {
    currentPage: 1,
    currentFilter: '',
    currentOrder: '',
    resultsVisible: false
  }
};

const sorts = Object.values(SORT_OPTIONS);

const selectProps = {
  selectName: 'product-order',
  selectTitle: ' Productos',
  classSelect: 'styles.filterbar__select',
  classTitle: 'styles.filterbar__optiontitle',
  classOption: 'styles.filterbar__option',
  optionValues: sorts
};

const handleMock = jest.fn();

const store = configureStore(initialState);

describe('Given the CustomSelect Component', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CustomSelect selectProps={selectProps} handleChange={handleMock} />
        </BrowserRouter>
      </Provider>
    );
  });
  test('then must be rendered (take a snapshot)', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <CustomSelect selectProps={selectProps} handleChange={handleMock} />
        </BrowserRouter>
      </Provider>
    );
    expect(
      asFragment(
        <CustomSelect selectProps={selectProps} handleChange={handleMock} />
      )
    ).toMatchSnapshot();
  });

  test('when the option change then call to handlechange', () => {
    const selectItem = screen.getByRole('combobox', {
      options: { name: 'product-order' }
    });

    const optionItems = screen.getAllByRole('option', { value: 'Precio' });

    // console.log(optionItems[1]); Is the option con value = price

    userEvent.selectOptions(selectItem, optionItems[1]);
    expect(handleMock).toHaveBeenCalled();
    expect(optionItems[1].value).toBe('Precio');
    expect(optionItems[1].selected).toBe(true);
  });
});
