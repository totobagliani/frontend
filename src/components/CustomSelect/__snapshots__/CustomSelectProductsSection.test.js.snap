// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the CustomSelect Component then must be rendered (take a snapshot) 1`] = `
<DocumentFragment>
  <select
    class="styles.filterbar__select"
    name="product-order"
  >
    <option
      class="styles.filterbar__optiontitle"
      selected=""
      value=""
    >
       Productos (2)
    </option>
    <option
      class="styles.filterbar__option"
      value="Precio"
    >
      Precio - 0
    </option>
    <option
      class="styles.filterbar__option"
      value="Nombre"
    >
      Nombre - 0
    </option>
    <option
      class="styles.filterbar__option"
      value="Sección"
    >
      Sección - 0
    </option>
  </select>
</DocumentFragment>
`;
